torch/_expanded_weights/expanded_weights_impl.py

```diff
@@ -0,0 +1,92 @@
+import torch
+import functools
+
+from typing import Callable, Dict
+
+HANDLED_FUNCTIONS: Dict[Callable, torch.autograd.Function] = {}
+
+def implements_per_sample_grads(torch_function):
+    @functools.wraps(torch_function)
+    def decorator(autograd_func):
+        HANDLED_FUNCTIONS[torch_function] = autograd_func
+        return autograd_func
+    return decorator
+
+# ExpandedWeight represents a weight (parameter) Tensor that has an expanded
+# batch dimension. Operations on the ExpandedWeight Tensor take advantage of
+# how the batch dimension is expanded by de-expanding the weight before
+# computation. A subsequent call to .backward() computes gradients for
+# ExpandedWeight. Those gradients are equivalent to per-sample-grads for the
+# unexpanded weight Tensors.
+#
+# ExpandedWeight has a fallback that does the forward + backward computation.
+# The backward computation is not optimized: it runs torch.autograd.grad in
+# a loop. To optimize the backward computation further, we must register
+# overrides for specific operators.
+#
+# This is a __torch_function__ object but it could have also been a Tensor Extension
+# with a dispatch key.
+class ExpandedWeight(torch.Tensor):
+    handled_functions = HANDLED_FUNCTIONS
+
+    # needed for conv2d default kwargs
+    conv_kwarg_options = ['stride', 'padding', 'dilation', 'groups']
+    conv_kwarg_defaults = {'stride': 1, 'padding': 0, 'dilation': 1, 'groups': 1}
+
+    def __new__(cls, orig_weight, batch_size):
+        ret = torch.Tensor._make_subclass(cls, orig_weight.detach(), orig_weight.requires_grad)
+        if not isinstance(orig_weight, torch.Tensor):
+            raise RuntimeError(f"Can only make Expanded Weights of Tensors, got {type(orig_weight).__name__}")
+        ret.batch_size = batch_size
+        ret.orig_weight = orig_weight
+        return ret
```

Comment:
mypy _really_ does not like this because "Tensor does not have attribute `batch_size`". Any thoughts? Otherwise I'll keep playing around with it and see if we can make something work

Reply:
Does a typing.cast() work here? (e.g. cast `ret` to ExpandedWeight. This is a no-op for everything aside from mypy.

Otherwise a mypy ignore (I forget the syntax) here is fine